### DISCUSSION TITLE
Keep it shell (consistency and terseness)

### DISCUSSION THREAD

**Discussion by morzel85** [nushell/nushell#7899] [General]
Nushell wants to make it easier to write and maintain non-trivial shell scripts with features like: types, immutability, scopes etc. Noble cause, no doubt! 

But... We mustn't forget that shell language should also be friendly for one-liners and just bashing (sorry) some things in terminal until you see the stuff you want.

Take this Nu as example:

`ls | where type == file | sort-by size`

Beautyful! Easy to type and easy to read. Self explanatory. 

Why it's so nice? Because commands implicitly operate on a structure returned by previous command. There is no need to provide a closure to filter values from `ls` nor we have to provide a comparator expression for sorting.

IMHO:
This sort of terseness should be preserved and encouraged. As many commands as possible should support it, optionally taking a closure for more involved predicates. 
Release 0.73 got the other way around for many commands. Where in the past we could write `all status == DONE` we now have to write `all {|x| $x.status == DONE}`.
I get that there might be technical reasons for wanting to do it, but this is an implementation detail escaping into the user interface. 
And by the way, I feel that pipe symbol used to surround closure arguments its also something that stems from implementation (Rust) and escaped into the interface. Pipe in shell is for pipelines of commands.

UX over DX.

P. S. I love Nushell and wish it well, so that's why I "waste" my Sunday on writing this :)

**Comment by kubouch** [comment]
Yeah, the `type == file` is called a "row condition". It would be possible to support row conditions but it would require all the related commands `all`, `any`, `skip while`, etc. to be parser keywords, that is, a part of the language syntax. So far we have opted for having only `where` as a parser keyword as in practice it is used all the time, unlike the other commands which are used much more rarely.

Unfortunately, supporting both a row condition and a closure in the same command is not possible because the syntax is ambiguous if you're trying to use a closure stored in a variable. That's why there are both `filter` and `where` commands.

You pointed out the tension we're dealing all the time: Are we more shell or a programming language? The goal of Nushell is to balance both, but this inevitably leads to tradeoffs on each side. Sometimes the tradeoffs are not optimal and I agree that having some kind of "closure shorthand" like a row condition but so far we haven't found an optimal solution.

In the meanwhile, you might try a poor-man's replacement:
```
> def all-cols [col: any, op: string, val: any] {
	if $op == '==' { 
		all {|x| ($x | get $col) == $val }
	} # ...add other cases as needed
}

> ls | all-cols type == dir
```

**Reply by morzel85** [reply]
Yes, finding a balance is difficult. I would personally vote for leaning more on the shell side... There are some solid choices on the programming side already (full static-dynamic spectrum), the shell side looks poorly in comparison.

**Reply by sophiajt** [reply]
Can you give some examples of leaning towards the shell side?

**Reply by morzel85** [reply]
> Can you give some examples of leaning towards the shell side?

Preferring ease of use of row conditions over referencing stored closure. Using such closure is a nice thing but relatively niche compared to accessing piped data directly so direct access should dictate the syntax and stored closure should fit around it.

"by leaning on shell side", I meant running any new syntax by such test:

- Does it make writing oneliners/interactive work easier or harder? Where easier means less characters and less punctuation.
- Does it make it easier to refer to data flowing from previous command? Let's pipe!

If syntax gets too heavy for rapid experimentation, one might as well switch from terminal to IDE, with all its niceties and start a new JS/Rust/Java/C#/Python project...

Here's some examples of Nu syntax I think is great considering the above points:

- I can write `[lorem ipsum]` and have a list of strings, I don't have to do `['lorem', 'ipsum']`
- I can write `ps | where cpu > 10` without providing predicate function.

**Comment by sophiajt** [comment]
Thanks for the feedback. This is one of those tricky ones we went back and forth on.

The change came from the UX/DX work in trying to make the shell nicer for people coming from other shells. One request we heard was to make redirections like `foo > bar.txt` work. In trying to do that, we saw that there was going to be ambiguity in the language. Here's one example:

```
ls | where $foo > bar.txt
```

Do you mean greater than, like`where ($foo > bar.txt)`,  or redirection with `where $foo` redirected to "bar.txt"? It's unclear. This is still true today, and I think we just pick that it means greater than rather than redirect. But for user commands, row conditions like what `where` uses looked like they could get pretty confusing.

I'm not sure exactly what the best design to land on is, but I'd be curious to keep iterating on this to see if we can find one.

**Reply by morzel85** [reply]
`>` as redirect looks tempting but maybe it will lead to too many ambiguities unless Nu switches to operators like `eq`, `gt`... And then people my want to add `<` for input...

A wild idea: would it help somehow if commands could always refer to implicit argument (iterator item, record etc.) that is distinguished from commands, operators and normal variables by having a prefix? Let's say `:` or `.` as its easy to type , eg:

`ls | where :name =~ 'lalala' and :size > $min`

I imagine it could help with autocomplete suggestions too.

**Reply by N4tus** [reply]
Isn't it currently out>, err>, and out+err> for redirection, that should be not that ambiguous?
